gazebo/common/SingletonT.hh

```diff
@@ -33,7 +33,7 @@ template <class T>
 class SingletonT
 {
   /// \brief Get an instance of the singleton
-  public: static T *Instance()
+  public: static T *Instance() GAZEBO_DEPRECATED(11.0)
```

Comment:
The basic idea is that this would print a warning if anyone was calling `SingletonT<Type>::Instance` in downstream code for one of the two reason:
* There was a class that inherited from `SingletonT<Type>`, without overriding `Instance`
* Even if an `Instance` in the derived class existed, the instance method of the base class was explicitly called for some reason, like ` `SingletonT<Type>::Instance()` 

Just preventing people from inheriting that class is not sufficient, as someone could call `SingletonT<Type>::Instance()` in two different compilation unit and/or libraries, with confusing behaviours.

Anyhow, I totally agree that this should be documented better in the Singleton class, I will do this.

Reply:
I tried to add some docs in https://github.com/gazebosim/gazebo-classic/pull/3269/commits/ff8269550c124a97b2ad73d90716703188565a43 . Given all the problems that this kind of class created, we could also consider to just document that the class is there for backward compatibility, and discourage future use.

Reply:
Thanks, the documentation looks good. I would also be in favor of discouraging future use. @scpeters are we okay with deprecating a function in a stable release? I'm leaning toward yes because this causes bugs that are hard to find. 

> Just preventing people from inheriting that class is not sufficient, as someone could call SingletonT<Type>::Instance()  in two different compilation unit and/or libraries, with confusing behaviours.

That's a fair point. If anyone calls `SingletonT<T>::Instance()` instead of `T::Instance` anywhere, we still risk getting duplicate symbols.

Reply:
yes, we can deprecate this

Reply:
Great, thanks!